### PR TITLE
Refactor for loop usage

### PR DESCRIPTION
**Pull request by thk123**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [n/a] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

Personally found the syntax for a regular incrementing for loop a bit cumbersome so provided a utility for the "typical" use.

### CODE REVIEW

**Comment by thk123** [comment]
Clang format issue fixed

File: src/util/std_code.cpp [hunk 1]
```diff
@@ -226,3 +226,22 @@ void code_outputt::check(const codet &code, const validation_modet vm)
   DATA_CHECK(
     vm, code.operands().size() >= 2, "output must have at least two operands");
 }
+
+code_fort code_fort::loop_from_to(
```
Comment (smowton) [inline src/util/std_code.cpp]: `from_index_bounds`? We already know it's a loop because it's a `code_fort`

File: src/util/std_code.cpp [hunk 2]
```diff
@@ -234,6 +234,8 @@ code_fort code_fort::loop_from_to(
   codet body,
   source_locationt location)
 {
+  PRECONDITION(start_index.type() == loop_index.type());
```
Comment (smowton) [inline src/util/std_code.cpp:237]: Commit message: compatabiltiy -> compatibility

**Review by smowton — APPROVED** [review]
A naming suggestion, otherwise lgtm

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fae1f65f8ce44f700616adc750c5a63fb5438307).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123405011

File: src/util/std_code.cpp [hunk 3]
```diff
@@ -226,3 +226,22 @@ void code_outputt::check(const codet &code, const validation_modet vm)
...
+{
+  side_effect_expr_assignt inc(
+    loop_index,
+    plus_exprt(loop_index, from_integer(1, loop_index.type())),
+    location);
+
+  return code_fort{
+    code_assignt{loop_index, std::move(start_index)},
+    binary_relation_exprt{loop_index, ID_lt, std::move(end_index)},
+    inc,
```
Comment (NathanJPhillips) [inline src/util/std_code.cpp]: `std::move`

File: jbmc/src/java_bytecode/remove_java_new.cpp [hunk 4]
```diff
@@ -299,11 +294,12 @@ goto_programt::targett remove_java_newt::lower_java_new_array(
...
-    const code_fort for_loop(
-      code_assignt(tmp_i, from_integer(0, tmp_i.type())),
-      binary_relation_exprt(tmp_i, ID_lt, to_multi_ary_expr(rhs).op0()),
-      std::move(inc),
-      std::move(for_body));
+    const code_fort for_loop = code_fort::loop_from_to(
+      from_integer(0, tmp_i.type()),
+      to_multi_ary_expr(rhs).op0(),
+      tmp_i,
+      for_body,
```
Comment (NathanJPhillips) [inline jbmc/src/java_bytecode/remove_java_new.cpp]: `std::move`

File: src/util/std_code.cpp (hunk 2 above)
Comment (NathanJPhillips) [reply src/util/std_code.cpp:237]: Or just squash into first commit!

File: jbmc/src/java_bytecode/java_object_factory.cpp [hunk 5]
```diff
@@ -1335,46 +1335,31 @@ static void array_loop_init_code(
...
-    array_element_from_pointer(array_init_symexpr, counter_expr);
+    const dereference_exprt element_at_counter =
+      array_element_from_pointer(array_init_symexpr, counter_expr);
 
-  assignments.append(element_generator(element_at_counter, element_type));
-  exprt java_one = from_integer(1, java_int_type());
-  code_assignt incr(counter_expr, plus_exprt(counter_expr, java_one));
+    loop_body.append(element_generator(element_at_counter, element_type));
+    return loop_body;
+  }();
```
Comment (NathanJPhillips) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: Say what???

**Review by NathanJPhillips — APPROVED** [review]

**Comment by thk123** [comment]
@NathanJPhillips @smowton comments [addressed](https://github.com/diffblue/cbmc/compare/fae1f65f8ce44f700616adc750c5a63fb5438307..af1c3ea3a5373e098a9b3d67eb253022ee4429d2) will merge when passing CI unless you want to re-check

**Comment by thk123** [comment]
Rebased on top of fixed doxygen

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0a84707ad325c7b3c30b54a8c9912df1e7f66b7c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124786728

**Pull request merged**